File: VB/E3647/ColumnLineHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Drawing
Imports DevExpress.XtraGrid.Columns

Namespace GridSample
	Friend Class ColumnLineHelper
		Private grid As GridControl
		Private view As GridView
		Private columnList As List(Of GridColumn)
		Public Property Columns() As List(Of GridColumn)
			Get
				Return columnList
			End Get
			Set(ByVal value As List(Of GridColumn))
				If columnList Is value Then
					Return
				End If
				columnList = value
				grid.Refresh()
			End Set
		End Property
		Private _hide As Boolean
		Public Property Hide() As Boolean
			Get
				Return _hide
			End Get
			Set(ByVal value As Boolean)
				If _hide = value Then
					Return
				End If
				_hide = value
				grid.Refresh()
			End Set
		End Property

		Public Sub New(ByVal gridView As GridView)
			view = gridView
			grid = gridView.GridControl
			AddHandler grid.PaintEx, AddressOf Grid_PaintEx
			columnList = New List(Of GridColumn)()
			Hide = True
		End Sub

		Private Sub Grid_PaintEx(ByVal sender As Object, ByVal e As PaintExEventArgs)
			Dim vi As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
			Dim standardPen As Pen = e.Cache.GetPen(view.PaintAppearance.Row.BackColor)
			Dim pen As Pen
			Dim p1 As Point = Point.Empty
			Dim p2 As Point = Point.Empty
			If Hide Then
				For Each ri As GridRowInfo In vi.RowsInfo
					If ri.IsGroupRow Then
						Continue For
					End If
					Dim p As Pen = standardPen
					If ri.RowHandle = view.FocusedRowHandle Then
						pen = e.Cache.GetPen(view.PaintAppearance.FocusedRow.BackColor)
					Else
						pen = standardPen
					End If
					For Each column As GridColumn In view.VisibleColumns
						If Not columnList.Contains(column) Then
							Continue For
						End If
						p1 = New Point(vi.ColumnsInfo(column).Bounds.Right - 1, ri.Bounds.Y)
						p2 = New Point(vi.ColumnsInfo(column).Bounds.Right - 1, ri.Bounds.Bottom - 2)
						e.Cache.DrawLine(pen, p1, p2)
					Next column
				Next ri
			End If
		End Sub

	End Class
End Namespace

File: CS/E3647/ColumnLineHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Grid;
using System.Drawing;
using DevExpress.XtraGrid.Columns;

namespace GridSample
{
    class ColumnLineHelper
    {
        GridControl grid;
        GridView view;
        List<GridColumn> columnList;
        public List<GridColumn> Columns
        {
            get { return columnList; }
            set
            {
                if (columnList == value)
                    return;
                columnList = value;
                grid.Refresh();
            }
        }
        bool _hide;
        public bool Hide
        {
            get { return _hide; }
            set {
                if (_hide == value)
                    return;
                _hide = value;
                grid.Refresh();
            }
        }

        public ColumnLineHelper(GridView gridView)
        {
            view = gridView;
            grid = gridView.GridControl;            
            grid.PaintEx += Grid_PaintEx;
            columnList = new List<GridColumn>();
            Hide = true;
        }

        private void Grid_PaintEx(object sender, PaintExEventArgs e) {
            GridViewInfo vi = view.GetViewInfo() as GridViewInfo;
            Pen standardPen = e.Cache.GetPen(view.PaintAppearance.Row.BackColor);
            Pen pen;
            Point p1 = Point.Empty;
            Point p2 = Point.Empty;
            if(Hide)
                foreach(GridRowInfo ri in vi.RowsInfo) {
                    if(ri.IsGroupRow)
                        continue;
                    Pen p = standardPen;
                    if(ri.RowHandle == view.FocusedRowHandle)
                        pen = e.Cache.GetPen(view.PaintAppearance.FocusedRow.BackColor);
                    else
                        pen = standardPen;
                    foreach(GridColumn column in view.VisibleColumns) {
                        if(!columnList.Contains(column)) continue;
                        p1 = new Point(vi.ColumnsInfo[column].Bounds.Right - 1, ri.Bounds.Y);
                        p2 = new Point(vi.ColumnsInfo[column].Bounds.Right - 1, ri.Bounds.Bottom - 2);
                        e.Cache.DrawLine(pen, p1, p2);
                    }
                }
        }
        
    }
}
